src/Native/System.Native/pal_networking.cpp

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "pal_config.h"
+#include "pal_networking.h"
+#include "pal_utilities.h"
+
+#include <arpa/inet.h>
+#include <assert.h>
+#include <functional>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#include <vector>
+
+const int NUM_BYTES_IN_IPV4_ADDRESS = 4;
+const int NUM_BYTES_IN_IPV6_ADDRESS = 16;
+
+static_assert(PAL_HOST_NOT_FOUND == HOST_NOT_FOUND, "");
+static_assert(PAL_TRY_AGAIN == TRY_AGAIN, "");
+static_assert(PAL_NO_RECOVERY == NO_RECOVERY, "");
+static_assert(PAL_NO_DATA == NO_DATA, "");
+static_assert(PAL_NO_ADDRESS == NO_ADDRESS, "");
+
+static void IpStringToAddressHelper(const char* address, const char* port, bool isIpV6, int32_t* err, std::function<void(const addrinfo& info)> lambda)
+{
+    assert(address != nullptr);
+    assert(err != nullptr);
+
+    addrinfo hint = {};
+    hint.ai_flags = AI_NUMERICHOST | AI_NUMERICSERV;
+    hint.ai_family = isIpV6 ? AF_INET6 : AF_INET;
+    hint.ai_socktype = 0;
+    hint.ai_protocol = 0;
+
+    addrinfo* info = nullptr;
+    int result = getaddrinfo(address, port, &hint, &info);
+    if (result == 0)
+    {
+        *err = 0;
+        lambda(*info);
+        freeaddrinfo(info);
+    }
+    else
+    {
+        *err = result;
+    }
+}
+
+static void ConvertByteArrayToV6SockAddrIn(sockaddr_in6& addr, const char* buffer, int32_t bufferLength)
+{
+    assert(bufferLength == NUM_BYTES_IN_IPV6_ADDRESS);
+    (void) bufferLength;
+
+    // IPv6 should have 4 32-bit ints, copy them to our buffer
+    const uint32_t* addressBuffer = reinterpret_cast<const uint32_t*>(buffer);
+    for (int i = 0; i < 4; i++)
+    {
+        // Some *nix distros use __in6_u and some use __u6_addr
+#if HAVE_IN6_U
+        addr.sin6_addr.__in6_u.__u6_addr32[i] = reinterpret_cast<const uint32_t*>(addressBuffer)[i];
+#else
+        addr.sin6_addr.__u6_addr.__u6_addr32[i] = reinterpret_cast<const uint32_t*>(addressBuffer)[i];
+#endif
+    }
+
+    // Mark that this is INET6
+    addr.sin6_family = AF_INET6;
+}
+
+static void ConvertV6SockAddrInToByteArray(char* buffer, int32_t bufferLength, const sockaddr_in6& addr)
+{
+    assert(bufferLength == NUM_BYTES_IN_IPV6_ADDRESS);
+    (void) bufferLength;
+
+    // IPv6 should have 4 32-bit ints, copy them to our buffer
+    uint32_t* output = reinterpret_cast<uint32_t*>(buffer);
+    for (int i = 0; i < 4; i++)
+    {
+        // Some *nix distros use __in6_u and some use __u6_addr
+#if HAVE_IN6_U
+        output[i] = addr.sin6_addr.__in6_u.__u6_addr32[i];
+#else
+        output[i] = addr.sin6_addr.__u6_addr.__u6_addr32[i];
+#endif
+    }
+}
+
+static void ConvertByteArrayToSockAddrIn(sockaddr_in& addr, const char* buffer, int32_t bufferLength)
+{
+    assert(bufferLength == NUM_BYTES_IN_IPV4_ADDRESS);
+    (void) bufferLength;
```

Comment:
This probably deserves a comment along the lines of "silence a compiler warning".